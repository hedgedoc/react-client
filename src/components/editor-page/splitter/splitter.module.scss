/*
 * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

.splitter {
  &.left {
    min-width: 200px;
  }

  &.separator {
    display: flex;
  }
}
