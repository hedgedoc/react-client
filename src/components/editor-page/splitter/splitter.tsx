/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import type { ReactElement } from 'react'
import React, { useCallback, useRef, useState } from 'react'
import { ShowIf } from '../../common/show-if/show-if'
import { SplitDivider } from './split-divider/split-divider'
import styles from './splitter.module.scss'
import { useAdjustedRelativeSplitValue } from './hooks/use-adjusted-relative-split-value'
import { useBindPointerMovementEventOnWindow } from '../../../hooks/common/use-bind-pointer-movement-event-on-window'

export interface SplitterProps {
  left?: ReactElement
  right?: ReactElement
  additionalContainerClassName?: string
  showLeft: boolean
  showRight: boolean
}

/**
 * Checks if the given {@link Event} is a {@link MouseEvent}.
 *
 * @param event the event to check
 * @return {@link true} if the given event is a {@link MouseEvent}
 */
const isMouseEvent = (event: Event): event is MouseEvent => {
  return (event as MouseEvent).buttons !== undefined
}

const isLeftMouseButtonClicked = (mouseEvent: MouseEvent): boolean => {
  return mouseEvent.buttons === 1
}

/**
 * Extracts the absolute horizontal position of the mouse or first touch point from the event.
 *
 * @param moveEvent The mouse or touch event that contains a pointer position
 * @return the extracted horizontal position or {@link undefined} if no position could be extracted
 */
const extractHorizontalPosition = (moveEvent: MouseEvent | TouchEvent): number | undefined => {
  if (isMouseEvent(moveEvent)) {
    return moveEvent.clientX
  } else {
    return moveEvent.touches[0]?.clientX
  }
}

/**
 * Creates a Left/Right splitter react component.
 *
 * @param additionalContainerClassName css classes that are added to the split container.
 * @param left the react component that should be shown on the left side.
 * @param right the react component that should be shown on the right side.
 * @param showLeft defines if the left component should be shown or hidden. Settings this prop will hide the component with css.
 * @param showRight defines if the right component should be shown or hidden. Settings this prop will hide the component with css.
 * @return the created component
 */
export const Splitter: React.FC<SplitterProps> = ({
  additionalContainerClassName,
  left,
  right,
  showLeft,
  showRight
}) => {
  const [relativeSplitValue, setRelativeSplitValue] = useState(50)
  const adjustedRelativeSplitValue = useAdjustedRelativeSplitValue(showLeft, showRight, relativeSplitValue)
  const resizingInProgress = useRef(false)
  const splitContainer = useRef<HTMLDivElement>(null)

  /**
   * Starts the splitter resizing.
   */
  const onStartResizing = useCallback(() => {
    resizingInProgress.current = true
  }, [])

  /**
   * Stops the splitter resizing.
   */
  const onStopResizing = useCallback(() => {
    if (resizingInProgress.current) {
      resizingInProgress.current = false
    }
  }, [])

  /**
   * Recalculates the panel split based on the absolute mouse/touch position.
   *
   * @param moveEvent is a {@link MouseEvent} or {@link TouchEvent} that got triggered.
   */
  const onMove = useCallback((moveEvent: MouseEvent | TouchEvent) => {
    if (!resizingInProgress.current || !splitContainer.current) {
      return
    }
    if (isMouseEvent(moveEvent) && !isLeftMouseButtonClicked(moveEvent)) {
      resizingInProgress.current = false
      moveEvent.preventDefault()
      return undefined
    }

    const horizontalPosition = extractHorizontalPosition(moveEvent)
    if (horizontalPosition === undefined) {
      return
    }
    const horizontalPositionInSplitContainer = horizontalPosition - splitContainer.current.offsetLeft
    const newRelativeSize = horizontalPositionInSplitContainer / splitContainer.current.clientWidth
    setRelativeSplitValue(newRelativeSize * 100)
    moveEvent.preventDefault()
  }, [])

  useBindPointerMovementEventOnWindow(onMove, onStopResizing)

  return (
    <div ref={splitContainer} className={`flex-fill flex-row d-flex ${additionalContainerClassName || ''}`}>
      <div
        className={`${styles['splitter']} ${styles['left']} ${!showLeft ? 'd-none' : ''}`}
        style={{ width: `calc(${adjustedRelativeSplitValue}% - 5px)` }}>
        {left}
      </div>
      <ShowIf condition={showLeft && showRight}>
        <div className={`${styles['splitter']} ${styles['separator']}`}>
          <SplitDivider onGrab={onStartResizing} />
        </div>
      </ShowIf>
      <div
        className={`${styles['splitter']}${!showRight ? ' d-none' : ''}`}
        style={{ width: `calc(100% - ${adjustedRelativeSplitValue}%)` }}>
        {right}
      </div>
    </div>
  )
}
