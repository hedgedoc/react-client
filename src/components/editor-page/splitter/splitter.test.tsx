/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import { fireEvent, render, screen } from '@testing-library/react'
import { Splitter } from './splitter'

describe('Splitter', () => {
  it('can render only the left pane', () => {
    const view = render(<Splitter showLeft={true} showRight={false} left={<>left</>} right={<>right</>} />)
    expect(view.container).toMatchSnapshot()
  })

  it('can render only the right pane', () => {
    const view = render(<Splitter showLeft={false} showRight={true} left={<>left</>} right={<>right</>} />)
    expect(view.container).toMatchSnapshot()
  })

  it('can render both panes', () => {
    const view = render(<Splitter showLeft={true} showRight={true} left={<>left</>} right={<>right</>} />)
    expect(view.container).toMatchSnapshot()
  })

  describe('resize', () => {
    beforeEach(() => {
      Object.defineProperty(window.HTMLDivElement.prototype, 'clientWidth', { value: 1920 })
      Object.defineProperty(window.HTMLDivElement.prototype, 'offsetLeft', { value: 0 })
    })

    it('can change size with mouse', async () => {
      const view = render(<Splitter showLeft={true} showRight={true} left={<>left</>} right={<>right</>} />)
      expect(view.container).toMatchSnapshot()
      const divider = await screen.findByTestId('splitter-divider')

      fireEvent.mouseDown(divider, {})
      fireEvent.mouseMove(window, { buttons: 1, clientX: 1920 })
      fireEvent.mouseUp(window)
      expect(view.container).toMatchSnapshot()

      fireEvent.mouseDown(divider, {})
      fireEvent.mouseMove(window, { buttons: 1, clientX: 0 })
      fireEvent.mouseUp(window)
      expect(view.container).toMatchSnapshot()

      fireEvent.mouseMove(window, { buttons: 1, clientX: 1920 })
      expect(view.container).toMatchSnapshot()
    })

    it('can change size with touch', async () => {
      const view = render(<Splitter showLeft={true} showRight={true} left={<>left</>} right={<>right</>} />)
      expect(view.container).toMatchSnapshot()
      const divider = await screen.findByTestId('splitter-divider')

      fireEvent.touchStart(divider, {})
      fireEvent.touchMove(window, { buttons: 1, clientX: 1920 })
      fireEvent.touchEnd(window)
      expect(view.container).toMatchSnapshot()

      fireEvent.touchStart(divider, {})
      fireEvent.touchMove(window, { buttons: 1, clientX: 0 })
      fireEvent.touchCancel(window)
      expect(view.container).toMatchSnapshot()

      fireEvent.touchMove(window, { buttons: 1, clientX: 1920 })
      expect(view.container).toMatchSnapshot()
    })
  })
})
