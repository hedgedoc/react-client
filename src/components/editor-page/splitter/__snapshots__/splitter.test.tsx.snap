// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Splitter can render both panes 1`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="splitter left "
      style="width: calc(50% - 5px);"
    >
      left
    </div>
    <div
      class="splitter separator"
    >
      <div
        class="split-divider"
        data-testid="splitter-divider"
      />
    </div>
    <div
      class="splitter"
      style="width: calc(100% - 50%);"
    >
      right
    </div>
  </div>
</div>
`;

exports[`Splitter can render only the left pane 1`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="splitter left "
      style="width: calc(100% - 5px);"
    >
      left
    </div>
    <div
      class="splitter d-none"
      style="width: calc(100% - 100%);"
    >
      right
    </div>
  </div>
</div>
`;

exports[`Splitter can render only the right pane 1`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="splitter left d-none"
      style="width: calc(0% - 5px);"
    >
      left
    </div>
    <div
      class="splitter"
      style="width: calc(100% - 0%);"
    >
      right
    </div>
  </div>
</div>
`;

exports[`Splitter resize can change size with mouse 1`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="splitter left "
      style="width: calc(50% - 5px);"
    >
      left
    </div>
    <div
      class="splitter separator"
    >
      <div
        class="split-divider"
        data-testid="splitter-divider"
      />
    </div>
    <div
      class="splitter"
      style="width: calc(100% - 50%);"
    >
      right
    </div>
  </div>
</div>
`;

exports[`Splitter resize can change size with mouse 2`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="splitter left "
      style="width: calc(100% - 5px);"
    >
      left
    </div>
    <div
      class="splitter separator"
    >
      <div
        class="split-divider"
        data-testid="splitter-divider"
      />
    </div>
    <div
      class="splitter"
      style="width: calc(100% - 100%);"
    >
      right
    </div>
  </div>
</div>
`;

exports[`Splitter resize can change size with mouse 3`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="splitter left "
      style="width: calc(0% - 5px);"
    >
      left
    </div>
    <div
      class="splitter separator"
    >
      <div
        class="split-divider"
        data-testid="splitter-divider"
      />
    </div>
    <div
      class="splitter"
      style="width: calc(100% - 0%);"
    >
      right
    </div>
  </div>
</div>
`;

exports[`Splitter resize can change size with mouse 4`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="splitter left "
      style="width: calc(0% - 5px);"
    >
      left
    </div>
    <div
      class="splitter separator"
    >
      <div
        class="split-divider"
        data-testid="splitter-divider"
      />
    </div>
    <div
      class="splitter"
      style="width: calc(100% - 0%);"
    >
      right
    </div>
  </div>
</div>
`;

exports[`Splitter resize can change size with touch 1`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="splitter left "
      style="width: calc(50% - 5px);"
    >
      left
    </div>
    <div
      class="splitter separator"
    >
      <div
        class="split-divider"
        data-testid="splitter-divider"
      />
    </div>
    <div
      class="splitter"
      style="width: calc(100% - 50%);"
    >
      right
    </div>
  </div>
</div>
`;

exports[`Splitter resize can change size with touch 2`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="splitter left "
      style="width: calc(100% - 5px);"
    >
      left
    </div>
    <div
      class="splitter separator"
    >
      <div
        class="split-divider"
        data-testid="splitter-divider"
      />
    </div>
    <div
      class="splitter"
      style="width: calc(100% - 100%);"
    >
      right
    </div>
  </div>
</div>
`;

exports[`Splitter resize can change size with touch 3`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="splitter left "
      style="width: calc(0% - 5px);"
    >
      left
    </div>
    <div
      class="splitter separator"
    >
      <div
        class="split-divider"
        data-testid="splitter-divider"
      />
    </div>
    <div
      class="splitter"
      style="width: calc(100% - 0%);"
    >
      right
    </div>
  </div>
</div>
`;

exports[`Splitter resize can change size with touch 4`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="splitter left "
      style="width: calc(0% - 5px);"
    >
      left
    </div>
    <div
      class="splitter separator"
    >
      <div
        class="split-divider"
        data-testid="splitter-divider"
      />
    </div>
    <div
      class="splitter"
      style="width: calc(100% - 0%);"
    >
      right
    </div>
  </div>
</div>
`;
